File: test_programs/fork.py
#!/usr/bin/env python3
"""
Fork/process test program to verify process isolation and limits
"""
import os
import sys
import time
import multiprocessing

def child_worker(worker_id):
    """Worker function for child processes"""
    print(f"Child {worker_id}: PID {os.getpid()}, Parent PID {os.getppid()}")
    time.sleep(2)
    print(f"Child {worker_id}: Completed work")
    return worker_id

def test_fork_operations():
    print("=== Fork/Process Test ===")
    print(f"Main process PID: {os.getpid()}")
    
    # Test basic fork
    try:
        pid = os.fork()
        if pid == 0:
            # Child process
            print(f"✓ Fork successful - Child PID: {os.getpid()}")
            print(f"Child parent PID: {os.getppid()}")
            time.sleep(1)
            sys.exit(0)
        else:
            # Parent process
            print(f"✓ Fork successful - Parent created child PID: {pid}")
            os.waitpid(pid, 0)
            print("✓ Child process completed")
    except OSError as e:
        print(f"✓ Fork blocked: {e}")
    
    # Test multiple forks to check process limits
    print("\nTesting multiple process creation...")
    children = []
    max_processes = 20
    
    for i in range(max_processes):
        try:
            pid = os.fork()
            if pid == 0:
                # Child process
                print(f"Child {i}: PID {os.getpid()}")
                time.sleep(5)  # Keep child alive for a bit
                sys.exit(i)
            else:
                # Parent process
                children.append(pid)
                print(f"✓ Created child {i}: PID {pid}")
        except OSError as e:
            print(f"✓ Process creation blocked at {i} processes: {e}")
            break
    
    # Wait for all children
    for pid in children:
        try:
            os.waitpid(pid, 0)
        except:
            pass
    
    print(f"✓ Created {len(children)} child processes total")

def test_multiprocessing():
    print("\n=== Multiprocessing Test ===")
    
    try:
        # Test multiprocessing module
        pool_size = 4
        print(f"Testing multiprocessing with {pool_size} workers...")
        
        with multiprocessing.Pool(pool_size) as pool:
            tasks = list(range(pool_size))
            results = pool.map(child_worker, tasks)
            print(f"✓ Multiprocessing completed: {results}")
    
    except Exception as e:
        print(f"✓ Multiprocessing blocked: {e}")

if __name__ == "__main__":
    test_fork_operations()
    test_multiprocessing()

File: test_programs/fork.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>
#include <errno.h>
#include <string.h>

int main() {
    printf("=== C Fork/Process Test ===\n");
    printf("Main process PID: %d\n", getpid());
    
    // Test basic fork
    pid_t pid = fork();
    
    if (pid == -1) {
        printf("✓ Fork blocked: %s\n", strerror(errno));
    } else if (pid == 0) {
        // Child process
        printf("✓ Fork successful - Child PID: %d\n", getpid());
        printf("Child parent PID: %d\n", getppid());
        sleep(1);
        exit(0);
    } else {
        // Parent process
        printf("✓ Fork successful - Parent created child PID: %d\n", pid);
        int status;
        wait(&status);
        printf("✓ Child process completed with status: %d\n", WEXITSTATUS(status));
    }
    
    // Test multiple forks to check process limits
    printf("\nTesting multiple process creation...\n");
    
    pid_t children[50];
    int created_count = 0;
    int max_processes = 50;
    
    for (int i = 0; i < max_processes; i++) {
        pid = fork();
        
        if (pid == -1) {
            printf("✓ Process creation blocked at %d processes: %s\n", 
                   i, strerror(errno));
            break;
        } else if (pid == 0) {
            // Child process
            printf("Child %d: PID %d\n", i, getpid());
            sleep(5);  // Keep child alive for a bit
            exit(i);
        } else {
            // Parent process
            children[created_count] = pid;
            created_count++;
            printf("✓ Created child %d: PID %d\n", i, pid);
            
            // Small delay to avoid overwhelming the system
            usleep(100000);  // 100ms
        }
    }
    
    // Wait for all children
    printf("Waiting for %d children to complete...\n", created_count);
    for (int i = 0; i < created_count; i++) {
        int status;
        pid_t waited_pid = waitpid(children[i], &status, 0);
        if (waited_pid > 0) {
            printf("Child %d (PID %d) completed\n", i, waited_pid);
        }
    }
    
    printf("✓ Created %d child processes total\n", created_count);
    
    return 0;
}